### PR TITLE
BREAK: get rid of `fqn_decorators.Decorator` and improve the type hinting

### PR DESCRIPTION
**Pull request by eigenein**

### CODE REVIEW

File: time_execution/timed.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+from typing import Any, Callable, Dict, List, Optional, Tuple, Type
+
+from fqn_decorators import get_fqn
+
+from time_execution import settings, write_metric
+
+SHORT_HOSTNAME = gethostname()
+
+
+class Timed(AbstractContextManager):
```
Comment (eigenein) [inline time_execution/timed.py:16]: This is extracted from the decorator code

File: time_execution/decorator.py [hunk 2]
```diff
@@ -1,91 +1,76 @@
...
+        # Backwards compatibility with `Decorator`.
+        wrapper.fqn = fqn  # type: ignore[attr-defined]
+        wrapper.get_fqn = lambda: wrapper.fqn  # type: ignore[attr-defined]
+        return cast(_F, wrapper)
+
+    return wrap(__wrapped) if __wrapped is not None else wrap
+
+
+# `time_execution` supports async out of the box.
+time_execution_async = time_execution
```
Comment (JeroennC) [inline time_execution/decorator.py:76]: Should we deprecate `time_execution_async`?

File: time_execution/decorator.py [hunk 3]
```diff
@@ -1,91 +1,76 @@
...
+                with Timed(wrapped=__wrapped, call_args=call_args, call_kwargs=call_kwargs, fqn=fqn, **kwargs) as timed:
+                    timed.result = __wrapped(*call_args, **call_kwargs)
+                    return timed.result
+
+        else:
+
+            @wraps(__wrapped)
+            async def wrapper(*call_args, **call_kwargs):
+                with Timed(wrapped=__wrapped, call_args=call_args, call_kwargs=call_kwargs, fqn=fqn, **kwargs) as timed:
+                    timed.result = await __wrapped(*call_args, **call_kwargs)
```
Comment (JeroennC) [inline time_execution/decorator.py:64]: Instead of setting the result in the `Timed` instance, we could yield the result to the `Timed` context manager, and return later

File: time_execution/timed.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    Both the sync and async decorators require the same logic around the wrapped function.
+    This context manager encapsulates the shared behaviour to avoid duplicating the code.
+    """
+
+    __slots__ = (
```
Comment (JeroennC) [inline time_execution/timed.py:20]: could maybe become a dataclass with slots=True

File: time_execution/decorator.py [hunk 5]
```diff
@@ -1,91 +1,76 @@
...
+
+    Args:
+        get_fqn: custom FQN getter (uses `fqn-decorators` by default)
+        extra_hooks: additional hooks (next to defined in the settings)
+        disable_default_hooks: if `True`, disable the hooks set by the settings
+    """
+
+
+def time_execution(__wrapped=None, get_fqn: Callable[[Any], str] = fqn_decorators.get_fqn, **kwargs):
+    from time_execution.timed import Timed
```
Comment (JeroennC) [inline time_execution/decorator.py:46]: Why is this necessary? Could you add a comment

Comment (eigenein) [reply time_execution/decorator.py:46]: What is exactly?

File: time_execution/decorator.py (hunk 3 above)
Comment (eigenein) [reply time_execution/decorator.py:64]: I don't really get it, any `yield` here would result in `wrapper()` becoming a generator (and so breaking the wrapped function)

File: time_execution/decorator.py (hunk 2 above)
Comment (eigenein) [reply time_execution/decorator.py:76]: Not sure, I don't mind it here actually

File: time_execution/timed.py (hunk 4 above)
Comment (eigenein) [reply time_execution/timed.py:20]: I could make it a `@dataclass`, alright, but `slots` is only available since 3.10

File: time_execution/decorator.py (hunk 5 above)
Comment (JeroennC) [reply time_execution/decorator.py:46]: The import inside the function call

File: time_execution/timed.py (hunk 4 above)
Comment (JeroennC) [reply time_execution/timed.py:20]: Ah.. then let's keep it this way for now

File: time_execution/decorator.py (hunk 5 above)
Comment (eigenein) [reply time_execution/decorator.py:46]: Ah, our ol' good friend Circular Dependency

File: time_execution/decorator.py (hunk 3 above)
Comment (JeroennC) [reply time_execution/decorator.py:64]: Good point, my initial thought was the wrong way around. Instead, I meant to use `Generator.send`, such that:
```
with Timed(wrapped=__wrapped, call_args=call_args, call_kwargs=call_kwargs, fqn=fqn, **kwargs) as timed:
                    response = await __wrapped(*call_args, **call_kwargs)
                    timed.send(response)
                    return response
```
and timed as a function (i know it's currently a class, not sure if that could also work) would be
```
@contextmanager
def timed(...):
   start_logic()
   result = yield
   end_logic()
```

Comment (eigenein) [reply time_execution/decorator.py:64]: I explored this option, and it would be even uglier, something like: `timed.gen.send(result)` (because a context manager is not a generator, and the original generator would be saved under the `.gen` attribute)

Comment (JeroennC) [reply time_execution/decorator.py:64]: Alright let's keep it this way then

File: time_execution/decorator.py (hunk 5 above)
Comment (JeroennC) [reply time_execution/decorator.py]: Okay fair enough, i try to always prevent circular dependencies but i assume you gave that a shot

**Review by JeroennC — DISMISSED** [review]

**Comment by eigenein** [comment]
@JeroennC Changed it to `BREAK`, the rest stays the same

**Review by JeroennC — APPROVED** [review]
